actors/evm/src/interpreter/precompiles.rs

```diff
@@ -93,7 +103,95 @@ fn read_right_pad<'a>(input: impl Into<Cow<'a, [u8]>>, len: usize) -> Cow<'a, [u
     input
 }
 
-// https://github.com/ethereum/go-ethereum/blob/25b35c97289a8db4753cdf5ab7f2b306ec71794d/core/vm/contracts.go#L165
+// --- Precompiles ---
+
+fn todo<RT: Runtime>(_: &RT, _: &[u8]) -> PrecompileResult {
+    todo!()
+}
+
+/// Read right padded BE encoded u64 ID address
+/// returns encoded CID or an empty array if actor not found
+fn get_actor_code_cid<RT: Runtime>(rt: &RT, input: &[u8]) -> PrecompileResult {
+    let id_bytes = read_right_pad(input, 8);
+    let id = u64::from_be_bytes(id_bytes.as_ref().try_into().unwrap());
+    Ok(rt.get_actor_code_cid(&id).unwrap_or_default().to_bytes())
+}
+
+/// Params:
+///
+/// | Param            | Value                     | Byte Length |
+/// |------------------|---------------------------|---|
+/// | randomness_type  | `Chain`(0) OR `Beacon`(1) | 1 |
+/// | RESERVED         | must be zeroes            | 7 |
+/// | personalization  | `i64` (LE encoded)        | 8 |
+/// | randomness_epoch | `i64` (LE encoded)        | 8 |
+/// | entropy_length   | `u64` (LE encoded)        | 8 |
+/// | entropy          | input\[32..] (right padded)| entropy_length |
+///
+/// Returns empty array if invalid randomness type
+/// Errors if unable to fetch randomness or bits are in reserved zone
+fn get_randomness<RT: Runtime>(rt: &RT, input: &[u8]) -> PrecompileResult {
+    // 1 + 7 (reserved) + 8 + 8 + 8 = 32
+    let word = read_right_pad(input, 32);
+
+    #[derive(num_derive::FromPrimitive)]
+    #[repr(u8)]
+    enum RandomnessType {
+        Chain = 0,
+        Beacon = 1,
+    }
+
+    let randomness_type = RandomnessType::from_u8(word[0]);
+
+    // 7 bytes reserved
+    if word[1..8] != [0u8; 7] {
+        return Err(PrecompileError::InvalidInput);
+    }
+
+    let personalization = i64::from_le_bytes(word[8..16].try_into().unwrap());
+    let rand_epoch = i64::from_le_bytes(word[16..24].try_into().unwrap());
+    let entropy_len = u64::from_le_bytes(word[24..32].try_into().unwrap());
```

Comment:
BE, not LE. We should stick with the conventions.

Reply:
Also... you're going to hate this but... I think we should stick with 32byte values because that's what we use elsewhere, and that's what's going to be fast to use from solidity.

Reply:
you are right and i hate it. may "u256 is a good general purpose value width" burn in the deepest depths of hell